index.rb

```diff
@@ -0,0 +1,145 @@
+# frozen_string_literal: true
+
+# rubocop:disable Metrics/CyclomaticComplexity
+# rubocop:disable Metrics/PerceivedComplexity
+# rubocop:disable Metrics/ModuleLength
+module Enumerable
+  def my_each
+    arr = to_a
+    return arr.to_enum unless block_given?
+
+    n = 0
+    while arr.length > n
+      yield(arr[n])
+      n += 1
+    end
+  end
+
+  def my_each_with_index
+    arr = to_a
+    return arr.to_enum unless block_given?
+
+    n = 0
+    while arr.length > n
+      yield(arr[n], n)
+      n += 1
+    end
+  end
+
+  def my_select
+    return to_enum unless block_given?
+
+    res = []
+    my_each do |v|
+      res << v if yield(v) == true
+    end
+    res
+  end
+
+  def my_all?(*args)
+    check = true
+    if block_given?
+      if args.empty?
+        my_each { |i| check = false unless yield(i) }
+      elsif args[0].is_a?(Class)
+        my_each { |i| check = false unless yield(i).class == args[0] }
+      elsif args[0].is_a?(Regexp)
+        my_each { |i| check = false unless (args[0]).match(yield(i)) }
+      else
+        my_each { |i| check = false unless yield(i) == args[0] }
+      end
+      return check
+    elsif args.empty?
+      my_all? { |obj| obj }
+    else
+      my_all?(args[0]) { |obj| obj }
+    end
+  end
+
+  def my_any?(*args)
+    check = false
+    if block_given?
+      if args.empty?
+        my_each { |i| check = true if yield(i) }
+      elsif args[0].is_a?(Class)
+        my_each { |i| check = true if yield(i).class == args[0] }
+      elsif args[0].is_a?(Regexp)
+        my_each { |i| check = true if yield(i).match(args[0]) }
+      else
+        my_each { |i| check = true if yield(i) == args[0] }
+      end
+      return check
+    elsif args.empty?
+      my_any? { |obj| obj }
+    else
+      my_any?(args[0]) { |obj| obj }
+    end
+  end
+
+  def my_none?(*args)
+    check = true
+    if block_given?
+      if args.empty?
+        my_each { |i| check = false if yield(i) }
+      elsif args[0].is_a?(Class)
+        my_each { |i| check = false if yield(i).class == args[0] }
+      elsif args[0].is_a?(Regexp)
+        my_each { |i| check = false if yield(i).match(args[0]) }
+      else
+        my_each { |i| check = false if yield(i) == args[0] }
+      end
+      return check
+    elsif args.empty?
+      my_none? { |obj| obj }
+    else
+      my_none?(args[0]) { |obj| obj }
+    end
+  end
+
+  def my_count(*args)
+    count = 0
+    if block_given?
+      if args.empty?
+        my_each { |i| count += 1 if yield(i) }
+      else
+        my_each { |i| count += 1 if yield(i) == args[0] }
+      end
+      return count
+    elsif args.empty?
+      my_count { |i| i }
+    else
+      my_count { |i| args[0] == i }
+    end
+  end
+
+  def my_map(&proc)
+    arr = to_a
+    return arr.to_enum unless block_given?
+
+    result = []
+    my_each { |i| result << proc.call(i) }
+    result
+  end
+
+  def my_inject(*args)
+    arr = to_a
+    acc = 0
+    if block_given?
```

Comment:
- [ ] my_inject combines all elements of enum by applying a binary operation, specified by a block


this method keeps failing

```
irb(main):145:0> [1,2,3,4,5,6].my_inject {|n,s| n*s}
=> 0
irb(main):146:0> [1,2,3,4,5,6].inject {|n,s| n*s}
=> 720
```